### PR TITLE
Add constructors

### PR DESCRIPTION
**Pull request by VikingScientist**

### CODE REVIEW

**Comment by VikingScientist** [comment]
This crashes with segmentation fault on one of the windows configurations only. Don't know why.

**Comment by VikingScientist** [comment]
Yes! Stupid fix, but it did fix it :)

File: lrspline/__init__.py [hunk 1]
```diff
@@ -80,6 +81,9 @@ def derivative(self, u, v, d=(1,1)):
         ])
         return retval.reshape(u.shape)
 
+    def __getitem__(self, idx):
```
Comment (TheBB) [inline lrspline/__init__.py:84]: There's two different `__getitem__` here that do different things. I guess the controlpoint one is more useful, so maybe delete this one.

File: lrspline/__init__.py [hunk 2]
```diff
@@ -389,16 +393,30 @@ def __mul__(self, x):
...
 class LRSplineSurface(LRSplineObject):
 
-    def __init__(self, arg=None):
-        if isinstance(arg, raw.LRSurface):
+    def __init__(self, *args):
+        if len(args) == 0:
+            w = raw.LRSurface()
+        elif isinstance(args[0], raw.LRSurface):
             w = arg
+        elif isinstance(args[0], io.IOBase):
```
Comment (TheBB) [inline lrspline/__init__.py]: This branch also supports strings and bytestrings, so I guess

```python
elif isinstance(args[0], (io.IOBase, str, bytes)):
```

File: lrspline/__init__.py [hunk 3]
```diff
@@ -389,16 +393,30 @@ def __mul__(self, x):
...
+            w = raw.LRSurface()
+            w.read(args[0])
+        elif len(args) == 2:
+            w = raw.LRSurface(args[0], args[1], 2, 2)
+        elif len(args) == 4:
+            w = raw.LRSurface(args[0], args[1], args[2], args[3])
+        elif len(args) == 6:
+            w = raw.LRSurface(args[0], args[1], args[2], args[3], args[4], args[5])
+        elif len(args) == 7:
+            w = raw.LRSurface(args[0], args[1], args[2], args[3], args[4], args[5], args[6])
```
Comment (TheBB) [inline lrspline/__init__.py:417]: Maybe these three paths can all be

```python
elif 4 <= len(args) <= 7:
    w = raw.LRSurface(*args)
```

File: lrspline/__init__.py [hunk 4]
```diff
@@ -461,18 +479,32 @@ def derivative(self, u, v, d=(1,1), iel=-1):
             return np.array(retval).reshape(u.shape)
         return self.w.point(u, v, nderivs, iEl=iel)[index]
 
+    def getBezierExtraction(self, iEl):
```
Comment (TheBB) [inline lrspline/__init__.py]: I'm trying to control names better in the Pythonic interface. I'd prefer `bezier_extraction`.

File: lrspline/__init__.py [hunk 5]
```diff
@@ -461,18 +479,32 @@ def derivative(self, u, v, d=(1,1), iel=-1):
...
+        if len(args) == 0:
+            w = raw.LRVolume()
+        elif isinstance(args[0], raw.LRVolume):
             w = arg
+        elif isinstance(args[0], io.IOBase):
+            w = raw.LRVolume()
+            w.read(args[0])
+        elif len(args) == 3: # only specify number of functions (n1,n2,n3)
+            w = raw.LRVolume(args[0], args[1], args[2], 2, 2, 2)
+        elif len(args) == 6: # specity n & p for 3 directions
```
Comment (TheBB) [inline lrspline/__init__.py:500]: Same here, these three paths are the same, so you can probably just do `if 6 <= len(args) <= 10`.

File: lrspline/__init__.py [hunk 6]
```diff
@@ -500,6 +532,9 @@ def evaluate(self, u, v, w, iel=-1):
             return retval.reshape(u.shape)
         return self.w.point(u, v, w, iEl=iel)
 
+    def getBezierExtraction(self, iEl):
```
Comment (TheBB) [inline lrspline/__init__.py]: Rename here as well.

File: lrspline/raw.pyx [hunk 7]
```diff
@@ -421,11 +438,29 @@ cdef class LRSplineObject:
     def setMaxAspectRatio(self, double r, bool aposteriori=True):
         self.w.setMaxAspectRatio(r, aposteriori)
 
-
 cdef class LRSurface(LRSplineObject):
 
-    def __cinit__(self, w=None):
-        self.w = new LRSplineSurface_()
+    def __cinit__(self, n1=None, n2=None, order_u=None, order_v=None, knot1=None, knot2=None, coef=None, dim=2):
+        cdef array.array knot1_arr;
```
Comment (TheBB) [inline lrspline/raw.pyx]: You don't need semicolons here.

File: lrspline/raw.pyx [hunk 8]
```diff
@@ -421,11 +438,29 @@ cdef class LRSplineObject:
...
+        cdef array.array knot2_arr;
+        cdef array.array coef_arr;
+        if knot1 is not None and knot2 is not None and coef is not None:
+            knot1_arr = array.array('d', knot1)
```
Comment (TheBB) [inline lrspline/raw.pyx:448]: I didn't try it but I guess it's also possible to do something like this:

```cpython
cdef vector[double] knot1_arr
knot1_arr = knot1
```

and then later, to get a pointer, do

```cpython
&knot1_arr[0]
```

But this is fine too.

**Review by TheBB — CHANGES_REQUESTED** [review]

File: lrspline/__init__.py (hunk 5 above)
Comment (VikingScientist) [reply lrspline/__init__.py:500]: The problem is when stupid users gives 5 input arguments. I decided that the best way was to just return the empty constructor in those cases. The reason for splitting it into multiple lines was for readability. While you can get away with:

``` python
if len(args) in (3,6,9,10): w = raw.LRVolume(*args)
```
is a one-line solution, I actually prefer a little longer more verbose implementation for documentation purposes.

File: lrspline/raw.pyx (hunk 7 above)
Comment (VikingScientist) [reply lrspline/raw.pyx]: But it is C-init... you have to get into the C mindset at that point 😉 

No, I'll fix it

File: lrspline/__init__.py (hunk 1 above)
Comment (VikingScientist) [reply lrspline/__init__.py:84]: There is only one `__getitem__` on LRSpline objects, right? The other one is for Basisfunctions. Returning controlpoints synergizes with what is in the `splipy`-library. Returning local knot vectors from Basisfunction objects might be more questionable, but it is what is done in the C-library and I guess the `BSplineBasis` from splipy as well, so it is consistent as well.

File: lrspline/__init__.py (hunk 2 above)
Comment (VikingScientist) [reply lrspline/__init__.py]: Was not aware of this. Will fix

File: lrspline/raw.pyx (hunk 8 above)
Comment (VikingScientist) [reply lrspline/raw.pyx:448]: This was picked up from a stackoverflow post which I can't seem to relocate. I don't know the exact reason it works, but it does so I dare not touch it. If you want to refactor it at a later point you are free to do this in a separate PR, but for now I think we'll keep it like this.

File: lrspline/__init__.py (hunk 1 above)
Comment (TheBB) [reply lrspline/__init__.py:84]: Fair, I couldn't see the whole context in github. :-P

**Review by TheBB — APPROVED** [review]

**Pull request merged**